From 135582654a6bd9ef045d729f3e5c30650910771e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jan=20Arve=20S=C3=A6ther?= <jan-arve.saether@qt.io>
Date: Tue, 18 Aug 2026 09:06:35 +0200
Subject: [PATCH] Don't dismiss popup while moving the mouse cursor with the
 touch pad

In order to dismiss a Popup on a touch screen/mouse, we checked
if the we pressed outside the popup, and then we closed the popup.

We checked for press by calling QPointerEvent::isBeginEvent(), and assumed
this meant that the point was pressed.

However, when the widget under the popup had Qt::WA_AcceptTouchEvents
set to true, and we **moved** the mouse cursor with a **Touch Pad** (on
macOS) outside of the popup, it would generate a QTouchEvent with
isBeginEvent() returning true on the inital touch.

The popup would then be dismissed, even if the user only moved the mouse
cursor with the touch pad.

We fix this by we adding an exception and don't dismiss the popup if the
deviceType() is TouchPad.

Note that the autotest doesn't really require to have a widget with
Qt::WA_AcceptTouchEvents under the popup, since we can synthesize
touchpad events ourselves from the autotest.

Fixes: QTBUG-147855
Pick-to: 6.11
Change-Id: Ifa4a1ca3e418fac614075d37fe6962a81980d7a3
Reviewed-by: Shawn Rutledge <shawn.rutledge@qt.io>
(cherry picked from commit c7fb2b0b7a3f6baec6decc8372c78c9d7ce2839e)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
Origin: upstream, https://invent.kde.org/qt/qt/qtdeclarative/-/commit/135582654a6bd9ef045d729f3e5c30650910771e
Forwarded: not-needed
Last-Update: 2026-10-29
---
 src/quicktemplates/qquickpopupwindow.cpp      |  6 ++-
 .../qquickpopup/tst_qquickpopup.cpp           | 43 +++++++++++++++++++
 2 files changed, 48 insertions(+), 1 deletion(-)

diff --git a/src/quicktemplates/qquickpopupwindow.cpp b/src/quicktemplates/qquickpopupwindow.cpp
index 4d3bc71346..192b367c5c 100644
--- a/src/quicktemplates/qquickpopupwindow.cpp
+++ b/src/quicktemplates/qquickpopupwindow.cpp
@@ -322,7 +322,11 @@ bool QQuickPopupWindowPrivate::filterPopupSpecialCases(QEvent *event)
             // of the shadow, targetMenu will still be nullptr.
             // On WASM in particular, it's possible for dialogs to receive the event, when clicking in the non-client area. Don't close in those cases.
             if (event->type() != QEvent::NonClientAreaMouseButtonPress && event->type() != QEvent::NonClientAreaMouseButtonDblClick) {
-                closePopups(QQuickPopup::CloseOnPressOutside | QQuickPopup::CloseOnPressOutsideParent);
+                const bool isTouchEventFromTouchPad = event->type() == QEvent::TouchBegin
+                    && pe->deviceType() == QInputDevice::DeviceType::TouchPad;
+                if (!isTouchEventFromTouchPad)
+                    closePopups(QQuickPopup::CloseOnPressOutside | QQuickPopup::CloseOnPressOutsideParent);
+
                 // A modal popup must consume the press event so that forwardToPopup() sees the event as handled
                 // so that it doesn't propagate to items behind the popup window (QTBUG-131786 etc.)
                 // A QTabletEvent in particular is not accepted by default
diff --git a/tests/auto/quickcontrols/qquickpopup/tst_qquickpopup.cpp b/tests/auto/quickcontrols/qquickpopup/tst_qquickpopup.cpp
index 3afb2387d6..49f8fddd82 100644
--- a/tests/auto/quickcontrols/qquickpopup/tst_qquickpopup.cpp
+++ b/tests/auto/quickcontrols/qquickpopup/tst_qquickpopup.cpp
@@ -88,6 +88,7 @@ private slots:
     void closePolicy_grabberInside();
     void closeOnRightClickOutside_data();
     void closeOnRightClickOutside();
+    void touchPadPressOutsideDoesNotClosePopupWindow();
     void activeFocusOnClose1();
     void activeFocusOnClose2();
     void activeFocusOnClose3();
@@ -843,6 +844,48 @@ void tst_QQuickPopup::closeOnRightClickOutside()
     QTRY_VERIFY(!popup->isVisible());
 }
 
+void tst_QQuickPopup::touchPadPressOutsideDoesNotClosePopupWindow()
+{
+    // QTBUG-147855: moving the mouse cursor with a touch pad (e.g. on macOS) can
+    // generate a QTouchEvent whose isBeginEvent() is true, if the item underneath
+    // has Qt::WA_AcceptTouchEvents set. Such an event must not be treated as a
+    // "press outside", and therefore must not dismiss the popup, even though it's
+    // a begin-event outside the popup and CloseOnPressOutside is in effect.
+    if (!arePopupWindowsSupported())
+        QSKIP("The platform doesn't support popup windows. Skipping test.");
+
+    QScopedPointer<QPointingDevice> touchPadDevice = QScopedPointer<QPointingDevice>(
+            QTest::createTouchDevice(QInputDevice::DeviceType::TouchPad));
+
+    QQuickControlsApplicationHelper helper(this, QStringLiteral("simplepopup.qml"));
+    QVERIFY2(helper.ready, helper.failureMessage());
+    QQuickWindow *window = helper.window;
+    window->show();
+
+    auto *popup = window->findChild<QQuickPopup *>();
+    QVERIFY(popup);
+    popup->setPopupType(QQuickPopup::Window);
+    // The default closePolicy() includes CloseOnPressOutside.
+
+    popup->open();
+    TRY_VERIFY_POPUP_OPENED(popup);
+    if (popup->popupType() != QQuickPopup::Window)
+        QSKIP("The popup did not resolve to a popup window on this platform. Skipping test.");
+
+    const QPoint outsidePoint(window->width() - 10, window->height() - 10);
+
+    // A touch press coming from a touch pad device must be ignored...
+    QTest::touchEvent(window, touchPadDevice.data()).press(0, outsidePoint);
+    QTest::touchEvent(window, touchPadDevice.data()).release(0, outsidePoint);
+    QVERIFY(popup->isOpened());
+
+    // ...whereas the same press coming from an actual touch screen should still
+    // close the popup, proving that the exception is specific to touch pad devices.
+    QTest::touchEvent(window, touchScreen.data()).press(0, outsidePoint);
+    QTest::touchEvent(window, touchScreen.data()).release(0, outsidePoint);
+    QTRY_VERIFY(!popup->isOpened());
+}
+
 void tst_QQuickPopup::activeFocusOnClose1()
 {
     SKIP_IF_NO_WINDOW_ACTIVATION;
